components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/dao/HybridPolicyDAOImpl.java

```diff
@@ -0,0 +1,253 @@
+import org.wso2.carbon.identity.entitlement.dto.PolicyDTO;
+import org.wso2.carbon.identity.entitlement.dto.PolicyStoreDTO;
+import org.wso2.carbon.identity.entitlement.policy.finder.AbstractPolicyFinderModule;
+import org.wso2.carbon.identity.entitlement.policy.finder.PolicyFinderModule;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.Set;
+
+/**
+ * HybridPolicyDAOImpl is a hybrid implementation of PolicyDAO. It uses both JDBC and Registry implementations to handle
+ * policy data. If the policy is already in the registry, it will be maintained there, including new versions.
+ * New policies will be persisted in the database.
+ */
+public class HybridPolicyDAOImpl extends AbstractPolicyFinderModule implements PolicyDAO {
+
+    private final JDBCPolicyDAOImpl jdbcPolicyDAO = new JDBCPolicyDAOImpl();
+    private final RegistryPolicyDAOImpl registryPolicyDAO = new RegistryPolicyDAOImpl();
+
+    @Override
+    public void init(Properties properties) {
+
+        jdbcPolicyDAO.init(properties);
+        registryPolicyDAO.init(properties);
+    }
+
+    @Override
+    public void addOrUpdatePolicy(PolicyDTO policy, boolean isPapUpdate) throws EntitlementException {
+
+        if (registryPolicyDAO.isPolicyExistsInPap(policy.getPolicyId())) {
+            registryPolicyDAO.addOrUpdatePolicy(policy, isPapUpdate);
+        } else {
+            jdbcPolicyDAO.addOrUpdatePolicy(policy, isPapUpdate);
+        }
+    }
+
+    @Override
+    public PolicyDTO getPAPPolicy(String policyId) throws EntitlementException {
+
+        PolicyDTO policyDTO = jdbcPolicyDAO.getPAPPolicy(policyId);
+        if (policyDTO == null) {
+            policyDTO = registryPolicyDAO.getPAPPolicy(policyId);
+        }
+        return policyDTO;
+    }
+
+    @Override
+    public List<PolicyDTO> getPAPPolicies(List<String> policyIds) throws EntitlementException {
+
+        List<PolicyDTO> policyDTOs = jdbcPolicyDAO.getPAPPolicies(policyIds);
+        List<PolicyDTO> regPolicyDTOs = registryPolicyDAO.getPAPPolicies(policyIds);
+        return EntitlementUtil.mergeAndRemoveDuplicates(policyDTOs, regPolicyDTOs);
+    }
+
+    @Override
+    public PolicyDTO getPolicy(String policyId, String version) throws EntitlementException {
+
+        PolicyDTO policyDTO;
+        try {
+            policyDTO = jdbcPolicyDAO.getPolicy(policyId, version);
+        } catch (EntitlementException e) {
+            policyDTO = registryPolicyDAO.getPolicy(policyId, version);
+        }
+        return policyDTO;
+    }
+
+    @Override
+    public String[] getVersions(String policyId) {
+
+        if (jdbcPolicyDAO.isPolicyExistsInPap(policyId)) {
+            return jdbcPolicyDAO.getVersions(policyId);
+        } else {
+            return registryPolicyDAO.getVersions(policyId);
+        }
+    }
+
+    @Override
+    public String getModuleName() {
+
+        return jdbcPolicyDAO.getModuleName();
```

Comment:
can consider to hard code this similar to `getSupportedSearchAttributesScheme()`